docs/beman-standard.md

```diff
@@ -0,0 +1,170 @@
+  recommendation can still conform to this standard. These entries are prefixed
+  by "RECOMMENDATION:".
+
+## License
+
+**[LICENSE.APPROVED]** REQUIREMENT: All Beman libraries must be licensed
+under an approved license. These are:
+
+1. [Apache License v2.0 with LLVM Exceptions](https://llvm.org/LICENSE.txt)
+2. [Boost Software License 1.0](https://www.boost.org/LICENSE_1_0.txt)
+3. [The MIT License](https://opensource.org/license/mit)
+
+**[LICENSE.BOOST]** RECOMMENDATION: All Beman libraries should be licensed under
+the [Apache License v2.0 with LLVM Exceptions](https://llvm.org/LICENSE.txt).
+
+**[LICENSE.CRITERIA]** REQUIREMENT: All approved licenses must meet the
+following requirements:
+
+1. Simple to read and understand.
+2. Permission without fee to copy, use and modify the software for any
+   use (commercial and non-commercial).
+3. Requires that the license appears on all copies of the software source code.
+4. Must not require that the license appears with executables or other binary
+   uses of the library.
+5. Must not require that the source code be available for execution or other
+   binary uses of the library.
+
+## General
+
+**[LIBRARY_NAMES]** REQUIREMENT: Beman libraries names begin with `Beman.`
+followed by an UpperCamelCase short name.
+
+Examples are Beman.SmartPointer and Beman.SenderReceiver.
+
+**[REPOSITORY_NAME]** RECOMMENDATION: The repository should be named after the
+library name excluding the `Beman.` prefix.
+
+For example, a Beman.SmartPointer library's repository should be named
+`SmartPointer`.
+
+## Top-level
+
+The top-level of a Beman library repository must consist of `CMakeLists.txt`,
+`LICENSE`, and `README.md` files.
+
+**[TOPLEVEL.CMAKE]** REQUIREMENT: There must be a `CMakeLists.txt` at the repository's root
+that builds and tests (via. CTest) the library.
+
+**[TOPLEVEL.LICENSE]** REQUIREMENT: There must be a `LICENSE` at the
+repository's root with the contents of an approved license that covers the
+contents of the repository.
+
+**[TOPLEVEL.README]** REQUIREMENT: There must be a markdown-formatted
+`README.md` at the repository's root that describes the library, explains how
+to build it, and links to further documentation.
+
+## `README.md`
+
+**[README.TITLE]** RECOMMENDATION: The `README.md` should begin with a level 1
+header with the name of the library optionally followed with a ":" and short
+description.
+
+For example:
+
+```markdown
+# Beman.SenderReceiver: Scalable Asychronous Program Building Blocks
+```
+
+**[README.PURPOSE]** RECOMMENDATION: Following the title, the `README.md` should
+contain a one- or two-paragraph summary describing the library's purpose.
+
+## Directory Layout
+
+**[DIRECTORY.INTERFACE_HEADERS]** REQUIREMENT: Header files that are part of the
+public interface must reside within the `include/Beman/\<ShortName\>/`
+directory.
+
+**[DIRECTORY.SOURCES]** RECOMMENDATION: Sources and headers not part of the
+public interface should reside in `src/`.
```

Comment:
it's a fairly usual convention in a header-only library to put the impl headers in a <shortname>/impl/ 

If we are going to put headers under src then we should use the pitchfork (from vector-of-bool) recommendations for what's called separated header placement:



```

In separated placement, there are two source directories, [§ 2.2 include/](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.include) and [§ 2.3 src/](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.src). The include/ directory is designated to contain the public headers of the library, while the src/ directory is designated to contain the compilable source code and private headers.In separated placement, there are two source directories, [§ 2.2 include/](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.include) and [§ 2.3 src/](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.src). The include/ directory is designated to contain the public headers of the library, while the src/ directory is designated to contain the compilable source code and private headers.

```

Reply:
I prefer the `libc++` convention.

Reply:
I'm mostly flexible. Fairly easy-to-meet technical asks:

* Includes in-repo must be identical to cross-project includes. Let's always include `foobar/core.hpp` and not sometimes include `../foobar/core.hpp` even though it's the same header.
* Let's prefix *every* include target with the library name. Yes to `foobar/core.hpp`. No thanks to `bits/core.hpp`. No thanks to `detail/foobar/core.hpp`. `libstdc++` gets this wrong for whatever reason. Effectively if project is named `libbits`, it sort of overlaps with `libstdc++` in its include target namespace. If `stdc++` put private headers in `libstdc++/internal/string.h` or something like that, it would be unambiguous what's going on.
* It should be hard to accidentally ship a header you didn't mean to. Probably this means shipped and unshipped headers live in entirely different subdirectories somehow.

Note that tools like include-what-you-use get confused regardless of the not-for-public-use convention, especially because it is a convention and not a standard. Eventually C++ needs an official way to declare what headers provide what symbols. Among the many reasons I'm also working on [CPS](https://github.com/cps-org/cps) is that it makes sense to eventually declare some of that sort of thing in a declarative JSON interface instead of by making up conventions per project.

Anyway, I'll go with consensus on this detail. I tend to put all of the source in one directory, personally. Careful naming of include targets covers all the ambiguity, but @camio is right that you can accidentally include something you didn't intend to ship that way. I personally find that the tooling to prevent that sort of bug is really lacking, but good enough test coverage of already packaged libraries (see my conan PR for example tests) would protect against any accidents there.

Reply:
I added [DIRECTORY.IMPLEMENTATION_HEADERS] which attempts to reflect the resolution we had at our last weekly meeting.